src/libfetchers/git-lfs-fetch.hh

```diff
@@ -0,0 +1,440 @@
+#include <iostream>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <regex>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+
+#include "serialise.hh"
+#include "processes.hh"
+#include "url.hh"
+
+namespace fs = std::filesystem;
+
+namespace nix {
+namespace lfs {
+
+// git-lfs metadata about a file
+struct Md
+{
+    std::string path; // fs path relative to repo root, no ./ prefix
+    std::string oid;  // git-lfs managed object id. you give this to the lfs server
+                      // for downloads
+    size_t size;      // in bytes
+};
+
+struct GitUrl
+{
+    std::string protocol;
+    std::string user;
+    std::string host;
+    std::string port;
+    std::string path;
+
+    std::string toHttp() const
+    {
+        if (protocol.empty() || host.empty()) {
+            return "";
+        }
+        std::string prefix = ((protocol == "ssh") ? "https" : protocol) + "://";
+        return prefix + host + (port.empty() ? "" : ":" + port) + "/" + path;
+    }
+
+    // [host, path]
+    std::pair<std::string, std::string> toSsh() const
+    {
+        if (host.empty()) {
+            return {"", ""};
+        }
+        std::string userPart = user.empty() ? "" : user + "@";
+        return {userPart + host, path};
+    }
+};
+
+struct Fetch
+{
+    // Reference to the repository
+    git_repository const * repo;
+
+    // Git commit being fetched
+    git_oid rev;
+
+    // from shelling out to ssh, used  for 2 subsequent fetches:
+    // list of URLs to fetch from, and fetching the data itself
+    std::string token = "";
+
+    // derived from git remote url
+    GitUrl gitUrl = GitUrl{};
+
+    Fetch(git_repository * repo, git_oid rev);
+    bool shouldFetch(const std::string & path) const;
+    void fetch(
+        const git_blob * pointerBlob,
+        const std::string & pointerFilePath,
+        Sink & sink,
+        std::function<void(uint64_t)> sizeCallback) const;
+    std::vector<nlohmann::json> fetchUrls(const std::vector<Md> & metadatas) const;
+};
+
+static size_t writeCallback(void * contents, size_t size, size_t nmemb, std::string * s)
+{
+    size_t newLength = size * nmemb;
+    s->append((char *) contents, newLength);
+    return newLength;
+}
+
+struct SinkCallbackData
+{
+    Sink * sink;
+    std::string_view sha256Expected;
+    HashSink hashSink;
+
+    SinkCallbackData(Sink * sink, std::string_view sha256)
+        : sink(sink)
+        , sha256Expected(sha256)
+        , hashSink(HashAlgorithm::SHA256)
+    {
+    }
+};
+
+static size_t sinkWriteCallback(void * contents, size_t size, size_t nmemb, SinkCallbackData * data)
+{
+    size_t totalSize = size * nmemb;
+    data->hashSink({(char *) contents, totalSize});
+    (*data->sink)({(char *) contents, totalSize});
+    return totalSize;
+}
+
+// if authHeader is "", downloadToSink assumes no auth is expected
+void downloadToSink(
+    const std::string & url, const std::string & authHeader, Sink & sink, std::string_view sha256Expected)
```

Comment:
This should use `FileTransfer::download()` (the sink variant). That way libfetcher doesn't need to reimplement curl support, and we get stuff like connection reuse.